File: packages/python/src/protocolbanks/http.py
"""
ProtocolBanks SDK - HTTP Client

HTTP 客户端封装，支持:
- 自动重试 (指数退避)
- 速率限制处理
- 请求/响应拦截
"""

from __future__ import annotations

import asyncio
from typing import Any, TypeVar

import httpx

from protocolbanks.errors import ProtocolBanksError, create_network_error
from protocolbanks.types import Environment, ErrorCodes, RetryConfig


T = TypeVar("T")

# ============================================================================
# Constants
# ============================================================================

BASE_URLS: dict[Environment, str] = {
    "production": "https://api.protocolbanks.com/v1",
    "sandbox": "https://sandbox-api.protocolbanks.com/v1",
    "testnet": "https://testnet-api.protocolbanks.com/v1",
}

DEFAULT_TIMEOUT = 30.0
DEFAULT_RETRY_CONFIG = RetryConfig(
    max_retries=3,
    initial_delay=1.0,
    max_delay=30.0,
    backoff_multiplier=2.0,
)


class HttpClient:
    """HTTP Client with retry and rate limiting"""

    def __init__(
        self,
        api_key: str,
        api_secret: str,
        environment: Environment = "production",
        base_url: str | None = None,
        timeout: float = DEFAULT_TIMEOUT,
        retry_config: RetryConfig | None = None,
    ):
        self._api_key = api_key
        self._api_secret = api_secret
        self._base_url = base_url or BASE_URLS.get(environment, BASE_URLS["production"])
        self._timeout = timeout
        self._retry_config = retry_config or DEFAULT_RETRY_CONFIG

        self._client = httpx.AsyncClient(
            base_url=self._base_url,
            timeout=timeout,
            headers=self._get_default_headers(),
        )

    def _get_default_headers(self) -> dict[str, str]:
        """Get default request headers"""
        return {
            "Content-Type": "application/json",
            "Accept": "application/json",
            "X-API-Key": self._api_key,
            "User-Agent": "ProtocolBanks-Python-SDK/0.1.0",
        }

    async def get(self, path: str, params: dict[str, Any] | None = None) -> Any:
        """Make GET request"""
        return await self._request("GET", path, params=params)

    async def post(self, path: str, data: dict[str, Any] | None = None) -> Any:
        """Make POST request"""
        return await self._request("POST", path, json=data)

    async def put(self, path: str, data: dict[str, Any] | None = None) -> Any:
        """Make PUT request"""
        return await self._request("PUT", path, json=data)

    async def delete(self, path: str) -> Any:
        """Make DELETE request"""
        return await self._request("DELETE", path)

    async def _request(
        self,
        method: str,
        path: str,
        params: dict[str, Any] | None = None,
        json: dict[str, Any] | None = None,
    ) -> Any:
        """Make HTTP request with retry"""
        last_error: Exception | None = None

        for attempt in range(self._retry_config.max_retries + 1):
            try:
                response = await self._client.request(
                    method=method,
                    url=path,
                    params=params,
                    json=json,
                )

                # Handle rate limiting
                if response.status_code == 429:
                    retry_after = int(response.headers.get("Retry-After", "60"))
                    if attempt < self._retry_config.max_retries:
                        await asyncio.sleep(retry_after)
                        continue
                    raise ProtocolBanksError(
                        code=ErrorCodes.RATE_LIMIT_EXCEEDED,
                        message="Rate limit exceeded",
                        retry_after=retry_after,
                        retryable=True,
                    )

                # Handle errors
                if response.status_code >= 400:
                    self._handle_error_response(response)

                # Return JSON response
                if response.content:
                    return response.json()
                return {}

            except httpx.TimeoutException:
                last_error = create_network_error(
                    "Request timed out",
                    ErrorCodes.NET_TIMEOUT,
                )
            except httpx.ConnectError:
                last_error = create_network_error(
                    "Connection failed",
                    ErrorCodes.NET_CONNECTION_FAILED,
                )
            except ProtocolBanksError:
                raise
            except Exception as e:
                last_error = ProtocolBanksError.from_exception(e)

            # Retry with backoff
            if attempt < self._retry_config.max_retries:
                delay = min(
                    self._retry_config.initial_delay
                    * (self._retry_config.backoff_multiplier**attempt),
                    self._retry_config.max_delay,
                )
                await asyncio.sleep(delay)

        # Raise last error
        if last_error:
            raise last_error

        raise create_network_error("Request failed after retries")

    def _handle_error_response(self, response: httpx.Response) -> None:
        """Handle error response"""
        try:
            data = response.json()
            error_code = data.get("code", ErrorCodes.NET_CONNECTION_FAILED)
            error_message = data.get("message", "Request failed")
            error_details = data.get("details")

            raise ProtocolBanksError(
                code=error_code,
                message=error_message,
                details=error_details,
                retryable=response.status_code >= 500,
            )
        except ValueError:
            # Non-JSON response
            raise ProtocolBanksError(
                code=ErrorCodes.NET_CONNECTION_FAILED,
                message=f"Request failed with status {response.status_code}",
                retryable=response.status_code >= 500,
            )

    async def close(self) -> None:
        """Close HTTP client"""
        await self._client.aclose()

    async def __aenter__(self) -> "HttpClient":
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.close()

File: packages/go/http.go
package protocolbanks

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// ============================================================================
// HTTP Client
// ============================================================================

// HTTPClient is an HTTP client for the ProtocolBanks API.
type HTTPClient struct {
	baseURL     string
	apiKey      string
	apiSecret   string
	httpClient  *http.Client
	retryConfig RetryConfig
}

// NewHTTPClient creates a new HTTP client.
func NewHTTPClient(config *Config) *HTTPClient {
	timeout := config.Timeout
	if timeout == 0 {
		timeout = 30 * time.Second
	}

	retryConfig := DefaultRetryConfig()
	if config.RetryConfig != nil {
		retryConfig = *config.RetryConfig
	}

	baseURL := config.BaseURL
	if baseURL == "" {
		baseURL = GetAPIBaseURL(config.Environment)
	}

	return &HTTPClient{
		baseURL:   baseURL,
		apiKey:    config.APIKey,
		apiSecret: config.APISecret,
		httpClient: &http.Client{
			Timeout: timeout,
		},
		retryConfig: retryConfig,
	}
}

// Get performs a GET request.
func (c *HTTPClient) Get(ctx context.Context, path string, result interface{}) error {
	return c.doRequest(ctx, http.MethodGet, path, nil, result)
}

// Post performs a POST request.
func (c *HTTPClient) Post(ctx context.Context, path string, body, result interface{}) error {
	return c.doRequest(ctx, http.MethodPost, path, body, result)
}

// Put performs a PUT request.
func (c *HTTPClient) Put(ctx context.Context, path string, body, result interface{}) error {
	return c.doRequest(ctx, http.MethodPut, path, body, result)
}

// Delete performs a DELETE request.
func (c *HTTPClient) Delete(ctx context.Context, path string, result interface{}) error {
	return c.doRequest(ctx, http.MethodDelete, path, nil, result)
}

func (c *HTTPClient) doRequest(ctx context.Context, method, path string, body, result interface{}) error {
	var lastErr error

	for attempt := 0; attempt <= c.retryConfig.MaxRetries; attempt++ {
		if attempt > 0 {
			delay := c.calculateDelay(attempt)
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(delay):
			}
		}

		err := c.executeRequest(ctx, method, path, body, result)
		if err == nil {
			return nil
		}

		lastErr = err

		// Check if error is retryable
		if sdkErr, ok := err.(*SDKError); ok {
			if !sdkErr.IsRetryable() {
				return err
			}
		}
	}

	return lastErr
}

func (c *HTTPClient) executeRequest(ctx context.Context, method, path string, body, result interface{}) error {
	url := c.baseURL + path

	var bodyReader io.Reader
	if body != nil {
		jsonBody, err := json.Marshal(body)
		if err != nil {
			return NewSDKError(ErrValidInvalidFormat, ErrorCategoryValid,
				"Failed to marshal request body")
		}
		bodyReader = bytes.NewReader(jsonBody)
	}

	req, err := http.NewRequestWithContext(ctx, method, url, bodyReader)
	if err != nil {
		return NewNetworkError("Failed to create request")
	}

	// Set headers
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("X-API-Key", c.apiKey)
	req.Header.Set("User-Agent", "protocolbanks-go/1.0.0")

	// Execute request
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return NewNetworkError("Request failed: " + err.Error())
	}
	defer resp.Body.Close()

	// Read response body
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return NewNetworkError("Failed to read response body")
	}

	// Handle error responses
	if resp.StatusCode >= 400 {
		return c.handleErrorResponse(resp.StatusCode, respBody)
	}

	// Parse successful response
	if result != nil && len(respBody) > 0 {
		if err := json.Unmarshal(respBody, result); err != nil {
			return NewSDKError(ErrValidInvalidFormat, ErrorCategoryValid,
				"Failed to parse response")
		}
	}

	return nil
}

func (c *HTTPClient) handleErrorResponse(statusCode int, body []byte) error {
	// Try to parse error response
	var errResp struct {
		Code    string `json:"code"`
		Message string `json:"message"`
	}

	if err := json.Unmarshal(body, &errResp); err == nil && errResp.Code != "" {
		category := c.categoryFromCode(errResp.Code)
		return NewSDKError(errResp.Code, category, errResp.Message)
	}

	// Generic error based on status code
	switch statusCode {
	case http.StatusUnauthorized:
		return NewSDKError(ErrAuthInvalidAPIKey, ErrorCategoryAuth, "Invalid API key")
	case http.StatusForbidden:
		return NewSDKError(ErrAuthInsufficientPermissions, ErrorCategoryAuth, "Insufficient permissions")
	case http.StatusNotFound:
		return NewSDKError(ErrValidInvalidFormat, ErrorCategoryValid, "Resource not found")
	case http.StatusTooManyRequests:
		return NewRateLimitError(time.Minute)
	case http.StatusInternalServerError:
		return NewNetworkError("Internal server error").WithRetryable(true)
	default:
		return NewNetworkError(fmt.Sprintf("Request failed with status %d", statusCode))
	}
}

func (c *HTTPClient) categoryFromCode(code string) ErrorCategory {
	if len(code) < 7 {
		return ErrorCategoryNet
	}

	// Extract category from code (e.g., PB_AUTH_001 -> AUTH)
	parts := code[3:] // Remove "PB_"
	for i, ch := range parts {
		if ch == '_' {
			cat := parts[:i]
			switch cat {
			case "AUTH":
				return ErrorCategoryAuth
			case "LINK":
				return ErrorCategoryLink
			case "X402":
				return ErrorCategoryX402
			case "BATCH":
				return ErrorCategoryBatch
			case "NET":
				return ErrorCategoryNet
			case "RATE":
				return ErrorCategoryRate
			case "VALID":
				return ErrorCategoryValid
			case "CRYPTO":
				return ErrorCategoryCrypto
			case "CHAIN":
				return ErrorCategoryChain
			}
			break
		}
	}

	return ErrorCategoryNet
}

func (c *HTTPClient) calculateDelay(attempt int) time.Duration {
	delay := c.retryConfig.InitialDelay
	for i := 1; i < attempt; i++ {
		delay = time.Duration(float64(delay) * c.retryConfig.BackoffMultiplier)
		if delay > c.retryConfig.MaxDelay {
			delay = c.retryConfig.MaxDelay
			break
		}
	}
	return delay
}

// Close closes the HTTP client.
func (c *HTTPClient) Close() error {
	c.httpClient.CloseIdleConnections()
	return nil
}
